Add pseudopotentials

mmezic opened the pull request:
This PR introduces the possibility of replacing several inner electrons of each atom with a pseudopotential. It currently supports the pseudopotentials from [Burkatzki, Filippi, Dolg 2008](http://aip.scitation.org/doi/10.1063/1.2987872) (`bfd`) and [Annaberdiyev et al. 2018](http://aip.scitation.org/doi/10.1063/1.5040472) (`ccECP`). The latter ones being the same as those used in the recent FermiNet paper ([Li et al. 2022](https://link.aps.org/doi/10.1103/PhysRevResearch.4.013021)).

The parameters of the pseudopotentials are parsed from the `PySCF` library and stored in the `Molecule` class. The numerical integration of nonlocal contributions is performed in a similar fashion as in Li et al. 2022, where a 12-point icosahedron quadrature was used. This makes the energy evaluation 12 times more demanding. However, the decrease in the number of explicitly treated electrons usually speeds up the overall computation.

### Example usage

```
mol = deepqmc.Molecule(
    coords=[[0, 0, 0], [3.679, 0, 0]],
    charges=[21, 8],
    charge=0,
    spin=1,
    unit='angstrom',
    pp_type='ccECP',
    pp_mask=[True, False],
)
```
Here, we define a ScO molecule with _ccECP_-type pseudopotential for the Sc atom. Due to `pp_mask`, O atom remains unaltered. This pseudopotential replaces 10 innermost electrons (i.e. Ne core) for Sc, reducing the number of active electrons from 29 to 19 in this case.

szbernat commented:
First of all, I really like the code in general, good job!

Here are a few minor comments I have:

1. Updating the docstrings by removing the `.jax` from the type specifications is a nice catch! Since we'll need to apply this fix on many other branches, it would be nice to have this as a separate PR. Maybe even squashed to a single commit?
2. Perhaps the function `load_pp_param` may be better placed in `pyscfext.py` instead of `molecule.py`?
3. Why do we use `int8` for `charges` [here](https://github.com/deepqmc/deepqmc/blob/96223598c38ab219f73d5208505d5104d31461d4/src/deepqmc/molecule.py#L158) instead of "default" `int32`?
4. In my mind, when using JAX/haiku, the name `params` is kind of reserved to mean the parameter dictionary of the model. Maybe we should consider renaming the parameters of the pseudopotentials to something like `pp_params`?
5. I think splitting [this large expression](https://github.com/deepqmc/deepqmc/blob/96223598c38ab219f73d5208505d5104d31461d4/src/deepqmc/physics.py#L69-L81) into smaller parts would improve readability. Maybe split into the terms of the equiations in the papers? Or assign the different parameters to temporary variables with shorter and more descriptive names instead of directly indexing `params`? Or anything else you can think of to reduce the complexity here..

zenoone commented on src/deepqmc/wf/baseline/pyscfext.py:
```diff
@@ -27,6 +32,12 @@ def pyscf_from_mol(mol, basis, cas=None, **kwargs):
         spin=mol.spin,
         cart=True,
         parse_arg=False,
+        ecp=dict(
```
I'd suggest this refactor:
```suggestion
        ecp={int(charge): mol.pp_type for charge in mol.charges[mol.pp_mask]}
```

mmezic replied:
This shorter code looks much better. Changed in 98009b312db1bc5afdac5e8bebe35794f77e2827

zenoone reviewed (commented):
First part of the review, physics.py will follow.

zenoone commented on src/deepqmc/physics.py:
```diff
@@ -46,6 +47,224 @@ def electronic_potential(rs):
+            [jnp.zeros_like(theta), jnp.ones_like(theta), jnp.zeros_like(theta)],
+            [-jnp.sin(theta), jnp.zeros_like(theta), jnp.cos(theta)],
+        ]
+    )
```
Definitions of rotation matrices could be moved to `utils.py`.

mmezic replied:
I agree. Moved in 69ac0f15209eb102c5f0b7dd20a23111ed019aa2

mmezic commented:
Thanks @szbernat for the comments.

> 1. Updating the docstrings by removing the `.jax` from the type specifications is a nice catch! Since we'll need to apply this fix on many other branches, it would be nice to have this as a separate PR. Maybe even squashed to a single commit?

You are right. I have made various unrelated changes apart from implementing pseudopotentials. Fortunately, Zeno had already split up the changes and left only the changes concerning pseudopotentials. Now, the commit history much cleaner. 

> 2. Perhaps the function `load_pp_param` may be better placed in `pyscfext.py` instead of `molecule.py`?

That is a good idea, I've moved the function to `pyscfext.py` in 526375eff52932f5573ecdf9b83237b82624f5bf.

> 3. Why do we use `int8` for `charges` [here]
(https://github.com/deepqmc/deepqmc/blob/96223598c38ab219f73d5208505d5104d31461d4/src/deepqmc/molecule.py#L158) instead of "default" `int32`?

I believe we'll never encounter a core with a charge larger that 127, so I thought that `int8` is enough and might be slightly faster. But the impact on memory and computational time is probably unobservable.

> 4. In my mind, when using JAX/haiku, the name `params` is kind of reserved to mean the parameter dictionary of the model. Maybe we should consider renaming the parameters of the pseudopotentials to something like `pp_params`?

Good point. I changed the name `params` to `loc_params` to be in line with the `nl_params` used in the non-local energy function.

> 5. I think splitting [this large expression](https://github.com/deepqmc/deepqmc/blob/96223598c38ab219f73d5208505d5104d31461d4/src/deepqmc/physics.py#L69-L81) into smaller parts would improve readability. Maybe split into the terms of the equiations in the papers? Or assign the different parameters to temporary variables with shorter and more descriptive names instead of directly indexing `params`? Or anything else you can think of to reduce the complexity here..

Another good point. This expression used to look like much simpler when I used just `*` operator, but then I needed to add an extra dimension, hence the `einsum`s. I must admit that I am also quite lost when looking at the expression now after the black formatter stretched it across many rows. I've tried to make it more readable in 9cbcc15fc717b7ffb709e757da0ce2513b409335

mmezic commented on src/deepqmc/molecule.py:
```diff
@@ -46,47 +102,103 @@ class Molecule:
+            pp_mask = [False] * len(charges)
+        if pp_mask is None:
+            pp_mask = [True] * len(charges)
+
```
The mask is actually set to `False` in the first `if` statement if no pseudopotential is used. Therefore the second `if` doesn't pass. I admit this part is not easily readable. Maybe swapping the order of both `if` statements could improve readability.

szbernat replied:
How about just using `elif` instead of the second if, making it explicit that it will only trigger if the first condition failed?

zenoone replied:
I see, my bad. I agree with Bernat that an `elif` statement could make the logic more transparent.

mmezic replied:
Changed to `elif` in 98009b312db1bc5afdac5e8bebe35794f77e2827

mmezic commented on src/deepqmc/molecule.py:
```diff
@@ -7,12 +7,62 @@
+        pp_nl_coef.append(jnp.swapaxes(jnp.array(data[1][i + 1][1][2]), -1, -2))
+    pp_nl_param = jnp.asarray(pp_nl_coef)
+    return n_core, pp_loc_param, pp_nl_param
+
```
The function was moved to `pyscfext.py` in 9cbcc15fc717b7ffb709e757da0ce2513b409335 and refactored according to your suggestions in 98009b312db1bc5afdac5e8bebe35794f77e2827 . I like your shorter code, but I had to fix an edge case in the `pp_nl_param` calculation, and sadly I wasn't able to keep it as a nice one-liner.

mmezic commented on src/deepqmc/molecule.py:
```diff
@@ -7,12 +7,62 @@
+    # The fixed padding is probably not the most efficient way,
+    # (it might cause some unnecessary multiplication by zero)
+    pad = 2
+
```
Sorry, this is an artifact from an older version. Luckily, I was able to restore the dynamic padding when I refactored the parsing of PP parameters in 98009b312db1bc5afdac5e8bebe35794f77e2827.

mmezic commented on src/deepqmc/molecule.py:
```diff
@@ -46,47 +102,103 @@ class Molecule:
+    # stores the parameters of local potential (loaded from [Burkatzki et al. 2007])
+    pp_loc_params: jnp.ndarray
+    # stores the parameters of non-local potential part
+    # (loaded from [Burkatzki et al. 2007])
```
Good point. Removed in 98009b312db1bc5afdac5e8bebe35794f77e2827

mmezic commented on src/deepqmc/molecule.py:
```diff
@@ -46,47 +102,103 @@ class Molecule:
+                pp_nl_params.append(jnp.asarray([]))
+        ns_core = jnp.array(ns_core)
+        pp_loc_params = jnp.array(pp_loc_params)
+
```
I think this is a great suggestion. I moved this part of the code to the parsing function (that I renamed `load_pp_params` -> `parse_pp_params`) in 98009b312db1bc5afdac5e8bebe35794f77e2827

mmezic commented on src/deepqmc/molecule.py:
```diff
@@ -46,47 +102,103 @@ class Molecule:
+            charges=1.0 * jnp.asarray(charges, dtype=jnp.int8),
             charge=charge,
             spin=spin,
+            pp_mask=pp_mask,
```
Applied in 98009b312db1bc5afdac5e8bebe35794f77e2827

mmezic commented on src/deepqmc/molecule.py:
```diff
@@ -46,47 +102,103 @@ class Molecule:
+            pp_mask = [True] * len(charges)
+
+        assert len(pp_mask) == len(charges), "Incompatible shape of 'pp_mask' given!"
+        pp_mask = jnp.array(pp_mask)
```
Applied in 98009b312db1bc5afdac5e8bebe35794f77e2827

mmezic commented on src/deepqmc/hamil/qc.py:
```diff
@@ -24,12 +26,15 @@ class MolecularHamiltonian(Hamiltonian):
+        self.elec_std = elec_std
 
-    def init_sample(self, rng, n, elec_std=1.0):
+    def init_sample(self, rng, n, elec_std=None):
```
Nice point. I had added the `elec_std` argument to the initialization of the hamiltonian due to better user accessibility. I guess this argument is no longer needed in `init_sample`, so I've removed it in 69ac0f15209eb102c5f0b7dd20a23111ed019aa2 . 

Exactly. We definitely need a better heuristic for setting the standard deviation. It currently scales as a square root of the atomic number (see https://github.com/deepqmc/deepqmc/blob/6a16c795bfc4a3e5a4ebc5ce9f51ab3591f58732/src/deepqmc/hamil/qc.py#L63). I think we should replace this square root by a more sophisticated function, but I don't know what'd be the appropriate choice here.

mmezic commented on src/deepqmc/hamil/qc.py:
```diff
@@ -71,11 +77,27 @@ def loc_ene(state, r, mol=self.mol):
+                    'hamil/V_el': Vs_el,
+                    'hamil/E_kin': Es_kin,
+                    'hamil/V_nuc': Vs_nuc,
+                }
```
Yes, `nuclear_potential` is a special case of `local_potential`. That is why I removed the `nuclear_potential` function completely, and now use only the `local_potential` function in  69ac0f15209eb102c5f0b7dd20a23111ed019aa2 . I have also included the simplifications you suggested

mmezic commented on src/deepqmc/physics.py:
```diff
@@ -46,6 +47,224 @@ def electronic_potential(rs):
+    y = rsin_theta * jnp.sin(phi)
+    z = r * jnp.cos(theta)
+    return jnp.array([x, y, z])
+
```
Moved in 69ac0f15209eb102c5f0b7dd20a23111ed019aa2

mmezic commented on src/deepqmc/physics.py:
```diff
@@ -46,6 +47,224 @@ def electronic_potential(rs):
+    # vmapping to be able to transform all 12 icosahedron points at the same time
+    transform_coordinates = jax.vmap(
+        transform_coordinates, in_axes=(None, None, None, -2, None)
+    )
```
This is because I need to vectorize the function along two extra dimensions. One dimension corresponds to the electron index and the other dimension covers the 12 icosahedron points. I believe this is not possible to be done with a single `vmap`, and I can't think of a more elegant way to write it

szbernat commented on tests/test_physics.py:
```diff
@@ -0,0 +1,24 @@
+        params, state, paulinet, rs = helpers.create_paulinet(
+            hamil=helpers.hamil(mol, elec_std=0.45)
+        )
+        print(rs)
```
Could you remove this `print` statement?

mmezic replied:
Sorry about that, I've removed it in dc4d99c430a903705e9a35fe97a084e6023353c2 .

zenoone commented on src/deepqmc/physics.py:
```diff
@@ -31,21 +32,208 @@ def pairwise_self_distance(coords, full=False):
+
+    # auxiliary function applying the rotation and translation to be vmapped
+    def transform_coordinates(norm, z_rot, y_rot, r, nucleus_position):
+        return norm * (z_rot @ y_rot @ r) + nucleus_position
```
This way of rotating the coordinates gives a fixes orientation of the quadrature points around the z axis of the icosahedron, based on its initialization. In principle there is some ambiguity, becaus a random rotation around the z-axis could be added befor the rotation around y. For a spherically symmetric system (e.g. an atom) fixing this ambiguity should not have a negative effect. However, it would be good to tests wether this can introduce a bias with respect to the molecular coordinates if a non spherically symmetric system is considerd (e.g. a diatomic molecule). If some  dependency on the relative initalization of the icosahedron wrt. the molecular coordinates can be observed I would suggest to add in a random rotation around z (like `(z_rot @ y_rot @ z_rot_random @ r)`).

mmezic replied:
This is a very good catch. I didn't test how significant the bias introduced by the preferred orientation is. But I believe that the bias could be present even for spherically symmetric systems, since the wavefunction need not have this symmetry. In 4dbef2bd8d0762aa563c1b28b2f04be6ff1727a2 I added the random rotation about the z-axis to be on the safe side. 

Note: I also had to add some non-trivial manipulations with PRNGkeys in the `fit_wf() `function to get a randomized rotation

zenoone replied:
Thanks! I guess you are right and in principle there could be a bias even if the Hamiltonian has some symmetry. The commit adding the random rotation looks good to me.

zenoone commented on src/deepqmc/physics.py:
```diff
@@ -31,21 +32,208 @@ def pairwise_self_distance(coords, full=False):
+    wf_vmapped = jax.vmap(wf, in_axes=(None, 0), out_axes=(0))
+
+    total_nl_potential = 0
+    for nucleus_index in range(mol.n_nuc):
```
Could it make sense to use a `jax.lax.fori_loop` here to improve the jitting?

mmezic replied:
The refactor from python `for` loop to `lax.fori_loop` is not as straightforward as it might seem. The problem is that different atoms have different numbers of pseudopotential terms. I have worked around this difficulty by using a differently sized array for each atom (I store these arrays in the list `mol.pp_nl_params`) and applying the `.sum()` method to each array. So the `lax.fori_loop` doesn't work here because it can only be applied to pure functions with fixed size arrays. I haven't found any clever workaround yet. I could make `mol.pp_nl_params` an array by padding it with zeros. But that would add a lot of unnecessary terms that have zero contribution. From my not very thorough testing, Jax doesn't optimise the zero terms, but the increase in computational cost from the additional pseudopotential terms is extremely small. So I think it makes sense to rewrite this using `lax.fori_loop` to improve the compile time whilst insignificantly increasing the calculation time. The code would also be in a better agreement with the code for local part of the pseudopotentials, where I used the `lax.fori_loop` with padding (additional zero terms). I'll first properly test the slowdown due to zero nonlocal terms before pushing the update.

mmezic replied:
Okay, after some more testing, I realized that the cost of the evaluating of pseudopotentials is FAR from significant. So it's indeed much better to use `lax.fori_loop`, because it really makes no sense to try to optimize this part at the cost of compile time. I have changed the `for` loop to `lax.fori_loop` in 549cd0eeaccacb727d4ea13fbb309cd887fb3e48
